From 04cb8cb30ea24710f85eb6bca2ec513ae886d562 Mon Sep 17 00:00:00 2001
From: Charles Chen <charlesccchen@google.com>
Date: Wed, 4 Feb 2026 05:45:21 +0000
Subject: [PATCH] Use setHideOverlayWindows instead of AppOpsManager

... to hide non-system overlay windows

Previously, DeviceAdminAdd added the check in onResume, and removed the
check in onPause. However it didn't stop overlay window showing again without
SYSTEM_ALERT_WINDOW permission when the restrictions are released.

This fix use Window#setHideOverlayWindows mechanism instead to prevent
any non-system overlay window shown on top of the window.

Test: presubmit
Bug: 463364901
Flag: EXEMPT bugfix
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:53fe03113a5e8fcf93cb3502c9de056956aaedd7
Merged-In: Idb5803fb87ac06f28be991d2dceeae600b29038a
Change-Id: Idb5803fb87ac06f28be991d2dceeae600b29038a
---
 .../enterprise/DeviceAdminAddActivity.java         |  8 ++++++++
 .../DeviceAdminAddHeaderPreferenceController.java  | 14 --------------
 2 files changed, 8 insertions(+), 14 deletions(-)

diff --git a/src/com/android/car/settings/enterprise/DeviceAdminAddActivity.java b/src/com/android/car/settings/enterprise/DeviceAdminAddActivity.java
index 69a3958dd..e8cd7ab9c 100644
--- a/src/com/android/car/settings/enterprise/DeviceAdminAddActivity.java
+++ b/src/com/android/car/settings/enterprise/DeviceAdminAddActivity.java
@@ -16,6 +16,8 @@
 
 package com.android.car.settings.enterprise;
 
+import android.os.Bundle;
+
 import androidx.fragment.app.Fragment;
 
 import com.android.car.settings.common.BaseCarSettingsActivity;
@@ -35,6 +37,12 @@ public final class DeviceAdminAddActivity extends BaseCarSettingsActivity {
     static final String EXTRA_CALLED_FROM_SUPPORT_DIALOG =
             "android.app.extra.CALLED_FROM_SUPPORT_DIALOG";
 
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        getWindow().setHideOverlayWindows(true);
+    }
+
     @Override
     protected Fragment getInitialFragment() {
         return new DeviceAdminAddFragment();
diff --git a/src/com/android/car/settings/enterprise/DeviceAdminAddHeaderPreferenceController.java b/src/com/android/car/settings/enterprise/DeviceAdminAddHeaderPreferenceController.java
index 67c2099f6..1e165a88a 100644
--- a/src/com/android/car/settings/enterprise/DeviceAdminAddHeaderPreferenceController.java
+++ b/src/com/android/car/settings/enterprise/DeviceAdminAddHeaderPreferenceController.java
@@ -17,13 +17,10 @@
 package com.android.car.settings.enterprise;
 
 import android.annotation.Nullable;
-import android.app.AppOpsManager;
 import android.car.drivingstate.CarUxRestrictions;
 import android.content.ComponentName;
 import android.content.Context;
 import android.graphics.drawable.Drawable;
-import android.os.Binder;
-import android.os.IBinder;
 import android.text.TextUtils;
 
 import androidx.preference.TwoStatePreference;
@@ -37,15 +34,11 @@ import com.android.internal.annotations.VisibleForTesting;
 public final class DeviceAdminAddHeaderPreferenceController
         extends BaseDeviceAdminAddPreferenceController<TwoStatePreference> {
 
-    private AppOpsManager mAppOps;
-    private final IBinder mToken = new Binder();
     private @Nullable ActivationListener mActivationListener;
 
     public DeviceAdminAddHeaderPreferenceController(Context context, String preferenceKey,
             FragmentController fragmentController, CarUxRestrictions uxRestrictions) {
         super(context, preferenceKey, fragmentController, uxRestrictions);
-
-        mAppOps = context.getSystemService(AppOpsManager.class);
     }
 
     DeviceAdminAddHeaderPreferenceController setActivationListener(ActivationListener listener) {
@@ -69,10 +62,6 @@ public final class DeviceAdminAddHeaderPreferenceController
     @VisibleForTesting
     void onResumeInternal(TwoStatePreference preference) {
         setCurrentStatus(preference);
-
-        // As long as we are running, don't let anyone overlay stuff on top of the screen.
-        mAppOps.setUserRestriction(AppOpsManager.OP_SYSTEM_ALERT_WINDOW, true, mToken);
-        mAppOps.setUserRestriction(AppOpsManager.OP_TOAST_WINDOW, true, mToken);
     }
 
     @Override
@@ -87,9 +76,6 @@ public final class DeviceAdminAddHeaderPreferenceController
     void onPauseInternal(TwoStatePreference preference) {
         // Disable the toggle button when paused, to prevent tapjacking.
         preference.setEnabled(false);
-
-        mAppOps.setUserRestriction(AppOpsManager.OP_SYSTEM_ALERT_WINDOW, false, mToken);
-        mAppOps.setUserRestriction(AppOpsManager.OP_TOAST_WINDOW, false, mToken);
     }
 
     @Override
-- 
2.55.0.897.gb25b4bd76c-goog

